src/game.cpp

```diff
@@ -4710,13 +4710,15 @@ template const Creature *game::critter_at<Creature>( const tripoint &, bool ) co
 template<typename T>
 shared_ptr_fast<T> game::shared_from( const T &critter )
 {
-    if( const shared_ptr_fast<monster> mon_ptr = critter_tracker->find( critter.pos() ) ) {
-        return std::dynamic_pointer_cast<T>( mon_ptr );
-    }
     if( static_cast<const Creature *>( &critter ) == static_cast<const Creature *>( &u ) ) {
         // u is not stored in a shared_ptr, but it won't go out of scope anyway
         return std::dynamic_pointer_cast<T>( u_shared_ptr );
     }
+    for( auto &cur_mon : critter_tracker->get_monsters_list() ) {
+        if( static_cast<const Creature *>( cur_mon.get() ) == static_cast<const Creature *>( &critter ) ) {
+            return std::dynamic_pointer_cast<T>( cur_mon );
+        }
+    }
```

Comment:
I don't think this is a good solution; it makes this function O(n) in the number of monsters, which could have a bad performance impact.

I think it would also work to instead replace the initial `if` block with:
```
    if( const shared_ptr_fast<monster> mon_ptr = critter_tracker->find( critter.pos() ) ) {
        if( shared_ptr<T> result = std::dynamic_pointer_cast<T>( mon_ptr ) ) {
            return result;
        }
    }
```
Then for this case (player riding monster) the cast would fail and the code would fall through to the next check and succeed in returning the player.